src/openexr.imageio/exrinput_c.cpp

```diff
@@ -0,0 +1,1576 @@
+        m_local_io.reset();
+        m_userdata.m_io = nullptr;
+        errorf("OpenEXR exception: %s", e.what());
+        return false;
+    } catch (...) {  // catch-all for edge cases or compiler bugs
+        m_local_io.reset();
+        m_userdata.m_io = nullptr;
+        errorf("OpenEXR exception: unknown");
+        return false;
+    }
+
+    m_userdata.m_img                = this;
+    exr_context_initializer_t cinit = EXR_DEFAULT_CONTEXT_INITIALIZER;
+
+    cinit.error_handler_fn = &oiio_exr_error_handler;
+    cinit.user_data        = &m_userdata;
+    // KDT TODO: We could change the above to not always create an
+    // ioproxy? Although if we do that, we'll have to change the logic
+    // in the error handler where right now we assume that only
+    // happens in the valid file check
+    if (m_userdata.m_io) {
+        cinit.read_fn = &oiio_exr_read_func;
+        cinit.size_fn = &oiio_exr_query_size_func;
+    }
+    // KDT TODO: Does OIIO offer malloc overrides?
+
+    exr_result_t rv = exr_start_read(&m_exr_context, name.c_str(), &cinit);
+    if (rv != EXR_ERR_SUCCESS) {
+        // the error handler would have already reported the error into us
+        m_local_io.reset();
+        m_userdata.m_io = nullptr;
+        return false;
+    }
+
+    rv = exr_get_count(m_exr_context, &m_nsubimages);
+    if (rv != EXR_ERR_SUCCESS) {
+        m_local_io.reset();
+        m_userdata.m_io = nullptr;
+        return false;
+    }
+
+    m_parts.resize(m_nsubimages);
+    m_subimage = -1;
+    m_miplevel = -1;
+
+    // Set up for the first subimage ("part"). This will trigger reading
+    // information about all the parts.
+    bool ok = seek_subimage(0, 0);
+    if (ok)
+        newspec = m_spec;
+    else
+        close();
+    return ok;
+}
+
+
+
+const ImageSpec&
+OpenEXRInput::init_part(int subimage, int miplevel)
+{
+    const PartInfo& part(m_parts[subimage]);
+    if (!part.initialized) {
+        // Only if this subimage hasn't yet been inventoried do we need
+        // to lock and seek, but that is only so we don't have to re-look values up
+        lock_guard lock(*this);
+        if (!part.initialized) {
+            if (!seek_subimage(subimage, miplevel)) {
+                errorf("Unable to initialize part");
+                return part.spec;
+            }
+        }
+    }
+
+    return part.spec;
+}
+
+
+
+bool
+OpenEXRInput::PartInfo::parse_header(OpenEXRInput* in, exr_context_t ctxt,
+                                     int subimage)
+{
+    bool ok = true;
+    if (initialized)
+        return ok;
+
+    ImageInput::lock_guard lock(*in);
+    spec = ImageSpec();
+
+    exr_result_t rv = exr_get_data_window(ctxt, subimage, &top_datawindow);
+    if (rv != EXR_ERR_SUCCESS)
+        return false;
+    rv = exr_get_display_window(ctxt, subimage, &top_displaywindow);
+    if (rv != EXR_ERR_SUCCESS)
+        return false;
+    spec.x           = top_datawindow.x_min;
+    spec.y           = top_datawindow.y_min;
```

Comment:
The diffs in this area make me wonder: Would it be a better choice if the C version uses structs in such a way that you say `top_datawindow.min.x` rather than `.x_min`, and then there's a lot more code that looks identical between the C and C++ versions? And you could even have handy functions (on the C++ side) that convert between the C and C++ equivalents of many of the imath classes.

Reply:
i debated that. You can certainly memory cast the structs, and I have unit tests to do so, and with your new constructors to imath, the vectors should just copy construct. I chose the "different" route to drive home that I'm not providing any of the other functionality in the c++ object, but maybe that is not the best way to indicate that

Reply:
It's your code and your call. But I think I would have chosen to make the analogs of the C++ structures organized such that in client code, the API calls obviously would be different between the C and C++ interfaces, but as much as possible of the secondary, more pedestrian code were identical for C and C++ in all the places that they could be  (like how you access struct members). It's not something I'm going to lose sleep over, but because it would change source compatibility, it's pretty much now or never to lock these choices in or rethink them.

BTW, I added the interop utilities only for Vec and Matrix types, but maybe it would be handy for Box, too.